### PR TITLE
Safekeeper: add timeline/tenant force delete HTTP endpoings

### PR DESCRIPTION
**Pull request by yeputons**
This implements the basic version of #895.

* There is no auth in Safekeeper HTTP at all currently, so simply calling `check_permission` (like in Pageserver) is not enough.
* There are no checks of Safekeeper still working with the data, as "still working" is burry now: a timeline may be "active" while there are no compute nodes and all data is propagated.
* However, there is an attempt to report at least some information about "active" timelines (even if inaccurate) back to the caller. I'm not sure if it's information useful, but it's there just in case. Maybe for some metrics or whatnot.

### CODE REVIEW

**Review by petuhovskiy — CHANGES_REQUESTED** [review]

File: safekeeper/src/http/routes.rs [hunk 1]
```diff
@@ -155,6 +238,14 @@ pub fn make_router(conf: SafeKeeperConf) -> RouterBuilder<hyper::Body, ApiError>
             timeline_status_handler,
         )
         .post("/v1/timeline", timeline_create_handler)
+        .delete(
+            "/v1/tenant/:tenant_id/timeline/:timeline_id/delete_force",
```
Comment (arssher) [inline safekeeper/src/http/routes.rs:242]: I'd drop `delete_force` part. It is strange to use delete method and repeat that in URL. And on force, currently we don't have a choice at all; if we have it later, we can specify in body the behaviour. Not very RESTy, but anyway.

I was also concerned about URL inconsistency with other methods, but that makes API closer to pageserver,  so ok, let's keep it this way and at some point make it fully identical.

File: safekeeper/src/http/routes.rs [hunk 2]
```diff
@@ -155,6 +238,14 @@ pub fn make_router(conf: SafeKeeperConf) -> RouterBuilder<hyper::Body, ApiError>
...
+            timeline_delete_force_handler,
+        )
+        .delete(
+            "/v1/tenant/:tenant_id/delete_all_force",
```
Comment (arssher) [inline safekeeper/src/http/routes.rs:246]: same here

File: libs/utils/src/http/request.rs [hunk 3]
```diff
@@ -31,3 +31,10 @@ pub fn parse_request_param<T: FromStr>(
         ))),
     }
 }
+
+pub async fn ensure_no_body(request: &mut Request<Body>) -> Result<(), ApiError> {
```
Comment (arssher) [inline libs/utils/src/http/request.rs:35]: do you have any tobacco?

File: safekeeper/src/http/routes.rs [hunk 4]
```diff
@@ -130,6 +131,88 @@ async fn timeline_create_handler(mut request: Request<Body>) -> Result<Response<
...
+async fn timeline_delete_force_handler(
+    mut request: Request<Body>,
+) -> Result<Response<Body>, ApiError> {
+    let zttid = ZTenantTimelineId::new(
+        parse_request_param(&request, "tenant_id")?,
+        parse_request_param(&request, "timeline_id")?,
+    );
+    ensure_no_body(&mut request).await?;
+    let conf = get_conf(&request);
+    let was_probably_active = GlobalTimelines::is_timeline_active(&zttid);
```
[the remark this answers is not in the document]
Comment (arssher) [reply safekeeper/src/http/routes.rs:156]: Right, we must delete map entry. Calling deactivate is also for now is useful.

File: safekeeper/src/http/routes.rs [hunk 5]
```diff
@@ -130,6 +131,88 @@ async fn timeline_create_handler(mut request: Request<Body>) -> Result<Response<
...
 
+#[derive(Serialize)]
+struct TimelineDeleteForceResponse {
+    pub dir_existed: bool,
+    pub was_probably_active: bool,
+}
+
+/// Removes the data directory for the timeline.
+///
+/// It does not try to stop any processing of the timeline; there is no such code at the time of writing.
```
[the remark this answers is not in the document]
Comment (arssher) [reply safekeeper/src/http/routes.rs:142]: Well, no. Timeline 'activeness' state machinery should be much better than it is now. We don't terminate walreceiver or walsender thread currently; basically the only (but important) purpose it currently serves is ensuring WAL commit_lsn reaches pageserver even if sync-safekeeper disconnects before that -- and then shut down callmemaybe.

File: safekeeper/src/lib.rs [hunk 6]
```diff
@@ -54,9 +54,12 @@ pub struct SafeKeeperConf {
 }
 
 impl SafeKeeperConf {
+    pub fn tenant_no_timeline_dir(&self, tenant_id: &ZTenantId) -> PathBuf {
```
Comment (arssher) [inline safekeeper/src/lib.rs:57]: simply `tenant_dir`?

**Review by arssher — CHANGES_REQUESTED** [review]

**Comment by yeputons** [comment]
Rebasing...

File: safekeeper/src/http/routes.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (yeputons) [reply safekeeper/src/http/routes.rs]: Done it, but note that even a simple HTTP GET request can potentially add an item to the map by implicit creation. That requires that the state can be loaded from the disk, though.

File: safekeeper/src/http/routes.rs [hunk 7]
```diff
@@ -130,6 +131,44 @@ async fn timeline_create_handler(mut request: Request<Body>) -> Result<Response<
     json_response(StatusCode::CREATED, ())
 }
 
+/// Deactivates the timeline and removes its data directory.
+///
+/// It does not try to stop any processing of the timeline; there is no such code at the time of writing.
+/// However, it tries to check whether the timeline was active and report it to caller just in case.
+/// Note that this information is inaccurate:
+/// 1. There is a race condition between checking the timeline for activity and actual directory deletion.
+/// 2. At the time of writing Safekeeper rarely marks a timeline inactive. E.g. disconnecting the compute node does nothing.
```
Comment (yeputons) [inline safekeeper/src/http/routes.rs:140]: Still relevant: I do not delete the directory under mutex, and the old (de)activation logic is still in place.

**Comment by yeputons** [comment]
Rebased, tests pass. I manually ensured that logs mention timeline deactivation, e.g. in test `test_delete_force`:

```
2022-05-12T21:15:34.151951Z  INFO {tid=15}:WAL acceptor{timeline=c8c39cbfe6876baf39b5ffae6b664223}: start receiving WAL since 0/16CD1B8
2022-05-12T21:15:34.165632Z  INFO {tid=17}: got query "IDENTIFY_SYSTEM"
2022-05-12T21:15:34.172847Z  INFO {tid=17}: got query "START_REPLICATION PHYSICAL 0/16CD1B8"
2022-05-12T21:15:34.173621Z  INFO {tid=17}:WAL sender{timeline=c8c39cbfe6876baf39b5ffae6b664223 pageserver_connstr=postgresql://no_user:@localhost:15000}: Start replication from 0/16CD1B8 till None
2022-05-12T21:15:34.938109Z  INFO {tid=15}:WAL acceptor{timeline=c8c39cbfe6876baf39b5ffae6b664223}: restart decoder from 0/0 to 0/16CD1B8
2022-05-12T21:15:35.145886Z  INFO deleting timeline 58d0f9b3b5154e1d8733dba1111fdde1-84fdb89cebf2caa5b6412134ee40f175
2022-05-12T21:15:35.146032Z  INFO timeline 84fdb89cebf2caa5b6412134ee40f175 is unsubscribed from callmemaybe to postgresql://no_user:@localhost:15000
2022-05-12T21:15:35.146289Z  INFO DELETE /v1/tenant/58d0f9b3b5154e1d8733dba1111fdde1/timeline/84fdb89cebf2caa5b6412134ee40f175 200 OK
2022-05-12T21:15:35.149809Z  INFO deleting timeline 58d0f9b3b5154e1d8733dba1111fdde1-84fdb89cebf2caa5b6412134ee40f175
2022-05-12T21:15:35.150074Z  INFO DELETE /v1/tenant/58d0f9b3b5154e1d8733dba1111fdde1/timeline/84fdb89cebf2caa5b6412134ee40f175 200 OK
2022-05-12T21:15:35.153847Z  INFO deleting timeline 58d0f9b3b5154e1d8733dba1111fdde1-d9a4f6bbf569d743cf83c9b3d45d9977
2022-05-12T21:15:35.158676Z  INFO DELETE /v1/tenant/58d0f9b3b5154e1d8733dba1111fdde1/timeline/d9a4f6bbf569d743cf83c9b3d45d9977 200 OK
2022-05-12T21:15:35.161881Z  INFO deleting timeline 58d0f9b3b5154e1d8733dba1111fdde1-00000000000000000000000000000000
2022-05-12T21:15:35.162059Z  INFO DELETE /v1/tenant/58d0f9b3b5154e1d8733dba1111fdde1/timeline/00000000000000000000000000000000 200 OK
2022-05-12T21:15:35.164292Z  INFO deleting all timelines for tenant 58d0f9b3b5154e1d8733dba1111fdde1
2022-05-12T21:15:35.164412Z  INFO timeline 7ffb084c773524c18fee3bbb3e7f8409 is unsubscribed from callmemaybe to postgresql://no_user:@localhost:15000
2022-05-12T21:15:35.168845Z  INFO DELETE /v1/tenant/58d0f9b3b5154e1d8733dba1111fdde1 200 OK
2022-05-12T21:15:35.172627Z  INFO deleting all timelines for tenant 58d0f9b3b5154e1d8733dba1111fdde1
2022-05-12T21:15:35.173583Z  INFO DELETE /v1/tenant/58d0f9b3b5154e1d8733dba1111fdde1 200 OK
2022-05-12T21:15:35.179149Z  INFO GET /v1/timeline/f76632e664b44c18b47c4b26d98f4fb1/c8c39cbfe6876baf39b5ffae6b664223 200 OK
2022-05-12T21:15:35.615155Z  INFO Got SIGQUIT. Terminating in immediate shutdown mode
```

**Review by arssher — APPROVED** [review]

**Review by petuhovskiy — APPROVED** [review]

**Comment by yeputons** [comment]
Fixed typo in last commit's message, will push without squashing once tests are green once again.

**Pull request merged**